File: analyze_horizon.py
"""
Horizon stress-test analysis: Measure how MAE grows with forecast horizon depth.

Two analyses:
1. CPM Stress-Test (Toto-only): How error grows across all rolling windows at different horizons.
   Demonstrates CPM's parallel decoding — inference time flat regardless of horizon.
2. Single-Origin Comparison (R + Toto): All models fitted once on 8,760h context, no re-scoring.
   Uses Toto window_id == 0 (2026-01-01 origin) to match R model origin exactly.

Segments:
  - 0-167h (1-week-ahead)
  - 168-335h (2-weeks-ahead)
  - 336-503h (3-weeks-ahead)
  - 504-671h (4-weeks-ahead)
"""
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from pathlib import Path

DATA_DIR = Path("data")

# Load actual test data for lookup beyond forecast window
actuals_df = pd.read_csv(DATA_DIR / "forecasts_r" / "actuals_hourly.csv")
actuals_df["datetime"] = pd.to_datetime(actuals_df["datetime"], format="mixed", utc=True)
actuals_df = actuals_df.set_index("datetime")
actuals_df.index = actuals_df.index.floor("h")

# Segment definitions (hour ranges)
SEGMENTS = {
    "0-167h\n(1 week)": (0, 168),
    "168-335h\n(2 weeks)": (168, 336),
    "336-503h\n(3 weeks)": (336, 504),
    "504-671h\n(4 weeks)": (504, 672),
}

SEGMENT_LABELS = ["0-167h (1 week)", "168-335h (2 weeks)",
                  "336-503h (3 weeks)", "504-671h (4 weeks)"]


# ---------------------------------------------------------------------------
# 1. CPM Stress-Test (Toto-only, all windows)
# ---------------------------------------------------------------------------

def load_toto_forecasts(horizon):
    """Load Toto forecasts for a given horizon.

    h160 is stored in data/forecasts_toto/toto_forecasts.csv (default).
    h320/h672 are in data/forecasts_toto_h{horizon}/toto_forecasts_h{horizon}.csv.
    """
    if horizon == 160:
        fpath = DATA_DIR / "forecasts_toto" / "toto_forecasts.csv"
    else:
        fpath = DATA_DIR / f"forecasts_toto_h{horizon}" / f"toto_forecasts_h{horizon}.csv"
    if not fpath.exists():
        return None
    df = pd.read_csv(fpath)
    df["datetime"] = pd.to_datetime(df["datetime"], utc=True)
    return df


def compute_toto_segment_mae(df, horizon):
    """Compute MAE for each horizon segment, averaged across all windows.

    Allows partial segments: if horizon=160, the 0-167h segment uses hours 0-159.
    Looks up actuals from test data for hours beyond the forecast window's target.
    """
    segment_errors = {name: [] for name in SEGMENTS}

    for wid, window in df.groupby("window_id"):
        window = window.sort_values("datetime")
        forecast_start = window["datetime"].iloc[0]

        for seg_name, (start, end) in SEGMENTS.items():
            if start >= horizon:
                continue

            seg_end = min(end, horizon, len(window))
            errors = []
            for i in range(start, seg_end):
                fc = window.iloc[i]["forecast_median"]
                act = window.iloc[i]["actual"]

                # Look up actual from test data if NaN
                if np.isnan(act):
                    dt = window.iloc[i]["datetime"]
                    try:
                        act = actuals_df.loc[dt, "actual"]
                    except (KeyError, TypeError):
                        continue

                if not np.isnan(act) and not np.isnan(fc):
                    errors.append(abs(act - fc))

            if errors:
                segment_errors[seg_name].append(np.mean(errors))

    return {name: np.mean(errs) if errs else np.nan
            for name, errs in segment_errors.items()}


def plot_cpm_stress_test():
    """Plot Toto-only error growth by horizon depth (CPM stress-test)."""
    results = {}

    for horizon in [160, 320, 672]:
        df = load_toto_forecasts(horizon)
        if df is not None:
            results[horizon] = compute_toto_segment_mae(df, horizon)

    fig, ax = plt.subplots(figsize=(10, 5))

    seg_names = list(SEGMENTS.keys())
    x_pos = range(len(seg_names))
    colors = {160: "#2563eb", 320: "#7c3aed", 672: "#dc2626"}
    widths = 0.25

    for horizon, color in colors.items():
        if horizon not in results:
            continue
        vals = [results[horizon].get(seg, np.nan) for seg in seg_names]
        offset = (horizon - 320) / 320 * widths
        bars = ax.bar([x + offset for x in x_pos], vals, widths,
                      label=f"Toto h={horizon}h", color=color, alpha=0.8)
        for bar, val in zip(bars, vals):
            if not np.isnan(val):
                ax.text(bar.get_x() + bar.get_width() / 2, bar.get_height() + 2,
                        f"{val:.0f}", ha="center", va="bottom", fontsize=8)

    ax.set_xticks(list(x_pos))
    ax.set_xticklabels(seg_names)
    ax.set_ylabel("MAE (vehicles/hour)")
    ax.set_title("Horizon Stress-Test: CPM Error Growth by Forecast Depth\n"
                  "(averaged across all rolling windows)")
    ax.legend(title="Total horizon")
    ax.grid(alpha=0.3, axis="y")
    plt.tight_layout()
    fig.savefig(DATA_DIR / "plots" / "horizon_stress_test.png", dpi=200, bbox_inches="tight")
    plt.close()

    print(f"  Saved: horizon_stress_test.png")
    return results


# ---------------------------------------------------------------------------
# 2. Single-Origin Comparison (R + Toto, same origin)
# ---------------------------------------------------------------------------

def load_r_forecasts(horizon):
    """Load R model forecasts for a given horizon."""
    fpath = DATA_DIR / "forecasts_r" / f"horizon_stress_h{horizon}.csv"
    if not fpath.exists():
        return None
    df = pd.read_csv(fpath)
    df["datetime"] = pd.to_datetime(df["datetime"], format="mixed", utc=True)
    return df


def compute_r_segment_mae(df, model_col, horizon):
    """Compute MAE for each horizon segment for an R model.

    Allows partial segments: if horizon=168, the 0-167h segment uses all 168 hours.
    """
    segment_errors = {}

    for seg_name, (start, end) in SEGMENTS.items():
        if start >= horizon:
            continue
        seg_end = min(end, horizon, len(df))
        fc = df[model_col].iloc[start:seg_end].values
        actual = df["actual"].iloc[start:seg_end].values
        mask = ~np.isnan(actual) & ~np.isnan(fc)
        if mask.any():
            segment_errors[seg_name] = np.mean(np.abs(actual[mask] - fc[mask]))

    return segment_errors


def compute_toto_single_origin_mae(horizon, window_id=7):
    """Compute MAE for Toto at a single origin (window_id=7, 2026-01-08).

    Window 7 starts at 2026-01-08, matching R models' origin.
    Context (2025-01-10 to 2025-12-31) includes Jan 1-2, 2025 holiday pattern.

    Looks up actuals from test data for hours beyond the forecast window's target.
    Allows partial segments when horizon < segment end.
    """
    df = load_toto_forecasts(horizon)
    if df is None:
        return {}

    # Filter to single origin
    window = df[df["window_id"] == window_id].sort_values("datetime")
    if window.empty:
        return {}

    forecast_start = window["datetime"].iloc[0]
    segment_errors = {}

    for seg_name, (start, end) in SEGMENTS.items():
        if start >= horizon:
            continue
        seg_end = min(end, horizon, len(window))

        errors = []
        for i in range(start, seg_end):
            fc = window.iloc[i]["forecast_median"]
            act = window.iloc[i]["actual"]

            # Look up actual from test data if NaN
            if np.isnan(act):
                dt = window.iloc[i]["datetime"]
                try:
                    act = actuals_df.loc[dt, "actual"]
                except (KeyError, TypeError):
                    continue

            if not np.isnan(act) and not np.isnan(fc):
                errors.append(abs(act - fc))

        if errors:
            segment_errors[seg_name] = np.mean(errors)

    return segment_errors


def compute_all_single_origin():
    """Compute segment MAE for all models at matching horizons, single origin."""
    # Map R horizons to Toto horizons (R uses 168/336/672, Toto uses 160/320/672)
    horizon_pairs = [
        ("r", 168, ("toto", 160)),
        ("r", 336, ("toto", 320)),
        ("r", 672, ("toto", 672)),
    ]

    results = {
        "SARIMA": {},
        "MSTL+ARIMA": {},
        "Naive": {},
        "Toto-2.5B": {},
    }

    r_model_cols = {
        "SARIMA": "sarima",
        "MSTL+ARIMA": "tbats",
        "Naive": "naive",
    }

    for r_label, r_horizon, (toto_label, toto_horizon) in horizon_pairs:
        # R models
        r_df = load_r_forecasts(r_horizon)
        if r_df is not None:
            for model_name, col in r_model_cols.items():
                seg_mae = compute_r_segment_mae(r_df, col, r_horizon)
                results[model_name].update(seg_mae)

        # Toto — single origin (window_id == 7, starts 2026-01-08)
        toto_seg_mae = compute_toto_single_origin_mae(toto_horizon, window_id=7)
        results["Toto-2.5B"].update(toto_seg_mae)

    return results


def plot_single_origin_comparison(results):
    """Plot R vs Toto error growth from single origin (Figure 5)."""
    fig, ax = plt.subplots(figsize=(10, 5))

    seg_names = list(SEGMENTS.keys())
    x_pos = np.arange(len(seg_names))
    width = 0.2

    models = {
        "SARIMA": {"color": "#e74c3c", "marker": "o"},
        "MSTL+ARIMA": {"color": "#3498db", "marker": "s"},
        "Naive": {"color": "#2ecc71", "marker": "^"},
        "Toto-2.5B": {"color": "#9b59b6", "marker": "D"},
    }

    offsets = [-1.5, -0.5, 0.5, 1.5]

    for idx, (model, style) in enumerate(models.items()):
        if model not in results:
            continue
        vals = []
        for seg in seg_names:
            if seg in results[model]:
                vals.append(results[model][seg])
            else:
                vals.append(np.nan)

        offset = offsets[idx] * width
        ax.plot([x + offset for x in x_pos], vals, style["marker"],
                label=model, color=style["color"], linewidth=2, markersize=7)
        for x, v in zip(x_pos, vals):
            if not np.isnan(v):
                ax.annotate(f"{v:.0f}", (x + offset, v),
                            textcoords="offset points", xytext=(0, 8),
                            ha="center", fontsize=7, color=style["color"])

    ax.set_xticks(list(x_pos))
    ax.set_xticklabels(seg_names)
    ax.set_ylabel("MAE (vehicles/hour)")
    ax.set_title(
        "Error Growth by Horizon Depth:\n"
        "Autoregressive (SARIMA, MSTL+ARIMA, Naive) vs CPM (Toto-2.5B)\n"
        "All models fitted once on 8,760h context, no re-scoring",
        y=1.05
    )
    ax.legend(title="Model", loc="upper left")
    ax.grid(alpha=0.3, axis="y")
    plt.tight_layout()
    fig.savefig(DATA_DIR / "plots" / "horizon_comparison.png", dpi=200, bbox_inches="tight")
    plt.close()

    print(f"  Saved: horizon_comparison.png")


def print_comparison_table(results):
    """Print MAE comparison table for paper."""
    seg_display = {
        "0-167h\n(1 week)": "1wk",
        "168-335h\n(2 weeks)": "2wk",
        "336-503h\n(3 weeks)": "3wk",
        "504-671h\n(4 weeks)": "4wk",
    }

    print("\nMAE by horizon segment (single origin, 2026-01-08):")
    print(f"{'Model':<15}", end="")
    for seg in seg_display:
        print(f"{seg_display[seg]:>10}", end="")
    print()

    for model in ["Toto-2.5B", "SARIMA", "MSTL+ARIMA", "Naive"]:
        if model not in results:
            continue
        print(f"{model:<15}", end="")
        for seg in seg_display:
            if seg in results[model]:
                print(f"{results[model][seg]:>10.1f}", end="")
            else:
                print(f"{'N/A':>10}", end="")
        print()


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    print("=" * 60)
    print("1. CPM Stress-Test (Toto-only, all rolling windows)")
    print("=" * 60)
    cpm_results = plot_cpm_stress_test()
    for horizon, segs in cpm_results.items():
        print(f"\n  Horizon={horizon}h:")
        for seg, mae in segs.items():
            seg_display = seg.replace("\n", " ")
            print(f"    {seg_display}: MAE={mae:.1f}" if not np.isnan(mae)
                  else f"    {seg_display}: N/A")

    print("\n" + "=" * 60)
    print("2. Single-Origin Comparison (R + Toto, 2026-01-08 origin)")
    print("=" * 60)
    comparison_results = compute_all_single_origin()
    plot_single_origin_comparison(comparison_results)
    print_comparison_table(comparison_results)

File: analyze_horizon.R
#!/usr/bin/env Rscript
# Horizon stress-test analysis: Measure how MAE grows with forecast horizon depth.
#
# Two analyses:
# 1. CPM Stress-Test (Toto-only): How error grows across all rolling windows at different horizons.
# 2. Single-Origin Comparison (R + Toto): All models fitted once on 8,760h context.
#    Uses Toto window_id == 7 (2026-01-08 origin) to match R model origin exactly.

suppressWarnings(suppressMessages({
  library(data.table)
  library(ggplot2)
}))

DATA_DIR <- "data"
PLOTS_DIR <- file.path(DATA_DIR, "plots")
dir.create(PLOTS_DIR, recursive = TRUE, showWarnings = FALSE)

# Load actual test data for lookup beyond forecast window
actuals_df <- fread(file.path(DATA_DIR, "forecasts_r", "actuals_hourly.csv"))
actuals_df[, datetime := as.POSIXct(datetime, tz = "UTC")]
setkey(actuals_df, datetime)

# Segment definitions (hour ranges, 0-indexed in forecast step but 1-based in indices)
# We will define the start and end as 1-based indices in R (1-168, 169-336, etc.)
SEGMENTS <- list(
  "0-167h\n(1 week)" = c(1, 168),
  "168-335h\n(2 weeks)" = c(169, 336),
  "336-503h\n(3 weeks)" = c(337, 504),
  "504-671h\n(4 weeks)" = c(505, 672)
)

# ---------------------------------------------------------------------------
# 1. CPM Stress-Test (Toto-only, all windows)
# ---------------------------------------------------------------------------

load_toto_forecasts <- function(horizon) {
  if (horizon == 160) {
    fpath <- file.path(DATA_DIR, "forecasts_toto", "toto_forecasts.csv")
  } else {
    fpath <- file.path(DATA_DIR, paste0("forecasts_toto_h", horizon), paste0("toto_forecasts_h", horizon, ".csv"))
  }
  if (!file.exists(fpath)) {
    return(NULL)
  }
  df <- fread(fpath)
  df[, datetime := as.POSIXct(datetime, tz = "UTC")]
  return(df)
}

compute_toto_segment_mae <- function(df, horizon) {
  # Storage for errors per segment
  segment_errors <- list()
  for (name in names(SEGMENTS)) {
    segment_errors[[name]] <- numeric()
  }
  
  # Group by window_id
  windows <- unique(df$window_id)
  
  for (wid in windows) {
    w <- df[window_id == wid][order(datetime)]
    
    for (seg_name in names(SEGMENTS)) {
      start <- SEGMENTS[[seg_name]][1]
      end <- SEGMENTS[[seg_name]][2]
      
      if (start > horizon) {
        next
      }
      
      seg_end <- min(end, horizon, nrow(w))
      errors <- numeric()
      
      for (i in start:seg_end) {
        fc <- w$forecast_median[i]
        act <- w$actual[i]
        
        # Look up actual from test data if NA
        if (is.na(act)) {
          dt <- w$datetime[i]
          lookup_val <- actuals_df[.(dt), actual]
          if (length(lookup_val) > 0) {
            act <- lookup_val
          }
        }
        
        if (!is.na(act) && !is.na(fc)) {
          errors <- c(errors, abs(act - fc))
        }
      }
      
      if (length(errors) > 0) {
        segment_errors[[seg_name]] <- c(segment_errors[[seg_name]], mean(errors))
      }
    }
  }
  
  results <- list()
  for (name in names(SEGMENTS)) {
    vals <- segment_errors[[name]]
    results[[name]] <- if (length(vals) > 0) mean(vals) else NA
  }
  return(results)
}

plot_cpm_stress_test <- function() {
  results <- list()
  
  for (horizon in c(160, 320, 672)) {
    df <- load_toto_forecasts(horizon)
    if (!is.null(df)) {
      results[[as.character(horizon)]] <- compute_toto_segment_mae(df, horizon)
    }
  }
  
  # Prepare data for plotting
  plot_data_list <- list()
  for (horizon_str in names(results)) {
    h_res <- results[[horizon_str]]
    for (seg_name in names(h_res)) {
      val <- h_res[[seg_name]]
      if (!is.na(val)) {
        plot_data_list[[length(plot_data_list) + 1]] <- data.table(
          Horizon = paste0("Toto h=", horizon_str, "h"),
          Segment = seg_name,
          MAE = val
        )
      }
    }
  }
  
  if (length(plot_data_list) == 0) {
    cat("No Toto forecasts found for stress test plotting.\n")
    return(results)
  }
  
  plot_df <- rbindlist(plot_data_list)
  # Fix factor levels for ordered segment plotting
  plot_df[, Segment := factor(Segment, levels = names(SEGMENTS))]
  
  colors_map <- c(
    "Toto h=160h" = "#2563eb",
    "Toto h=320h" = "#7c3aed",
    "Toto h=672h" = "#dc2626"
  )
  
  p1 <- ggplot(plot_df, aes(x = Segment, y = MAE, fill = Horizon)) +
    geom_bar(stat = "identity", position = position_dodge(width = 0.8), width = 0.7, alpha = 0.85) +
    geom_text(aes(label = sprintf("%.0f", MAE)), 
              position = position_dodge(width = 0.8), vjust = -0.3, size = 3) +
    scale_fill_manual(values = colors_map) +
    labs(
      title = "Horizon Stress-Test: CPM Error Growth by Forecast Depth\n(averaged across all rolling windows)",
      x = "Forecast Horizon Depth",
      y = "MAE (vehicles/hour)",
      fill = "Total Horizon"
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(face = "bold", size = 11, hjust = 0.5),
      legend.position = "bottom"
    )
  
  ggsave(file.path(PLOTS_DIR, "horizon_stress_test.png"), plot = p1, width = 9, height = 5.5, dpi = 150)
  cat("  Saved: horizon_stress_test.png\n")
  
  return(results)
}

# ---------------------------------------------------------------------------
# 2. Single-Origin Comparison (R + Toto, same origin)
# ---------------------------------------------------------------------------

load_r_forecasts <- function(horizon) {
  fpath <- file.path(DATA_DIR, "forecasts_r", paste0("horizon_stress_h", horizon, ".csv"))
  if (!file.exists(fpath)) {
    return(NULL)
  }
  df <- fread(fpath)
  df[, datetime := as.POSIXct(datetime, tz = "UTC")]
  return(df)
}

compute_r_segment_mae <- function(df, model_col, horizon) {
  segment_errors <- list()
  
  for (seg_name in names(SEGMENTS)) {
    start <- SEGMENTS[[seg_name]][1]
    end <- SEGMENTS[[seg_name]][2]
    
    if (start > horizon) {
      next
    }
    
    seg_end <- min(end, horizon, nrow(df))
    
    # 1-based indexing in R
    fc <- df[[model_col]][start:seg_end]
    actual <- df$actual[start:seg_end]
    
    mask <- !is.na(actual) & !is.na(fc)
    if (any(mask)) {
      segment_errors[[seg_name]] <- mean(abs(actual[mask] - fc[mask]))
    }
  }
  
  return(segment_errors)
}

compute_toto_single_origin_mae <- function(horizon, target_window_id = 7) {
  df <- load_toto_forecasts(horizon)
  if (is.null(df)) {
    return(list())
  }
  
  # Filter to single origin
  w <- df[window_id == target_window_id][order(datetime)]
  if (nrow(w) == 0) {
    return(list())
  }
  
  segment_errors <- list()
  
  for (seg_name in names(SEGMENTS)) {
    start <- SEGMENTS[[seg_name]][1]
    end <- SEGMENTS[[seg_name]][2]
    
    if (start > horizon) {
      next
    }
    
    seg_end <- min(end, horizon, nrow(w))
    errors <- numeric()
    
    for (i in start:seg_end) {
      fc <- w$forecast_median[i]
      act <- w$actual[i]
      
      # Look up actual from test data if NA
      if (is.na(act)) {
        dt <- w$datetime[i]
        lookup_val <- actuals_df[.(dt), actual]
        if (length(lookup_val) > 0) {
          act <- lookup_val
        }
      }
      
      if (!is.na(act) && !is.na(fc)) {
        errors <- c(errors, abs(act - fc))
      }
    }
    
    if (length(errors) > 0) {
      segment_errors[[seg_name]] <- mean(errors)
    }
  }
  
  return(segment_errors)
}

compute_all_single_origin <- function() {
  # Map R horizons to Toto horizons (R uses 168/336/672, Toto uses 160/320/672)
  horizon_pairs <- list(
    list(r = 168, toto = 160),
    list(r = 336, toto = 320),
    list(r = 672, toto = 672)
  )
  
  results <- list(
    SARIMA = list(),
    `MSTL+ARIMA` = list(),
    Naive = list(),
    `Toto-2.5B` = list()
  )
  
  r_model_cols <- c(
    SARIMA = "sarima",
    `MSTL+ARIMA` = "tbats",
    Naive = "naive"
  )
  
  for (pair in horizon_pairs) {
    r_horizon <- pair$r
    toto_horizon <- pair$toto
    
    # R models
    r_df <- load_r_forecasts(r_horizon)
    if (!is.null(r_df)) {
      for (model_name in names(r_model_cols)) {
        col <- r_model_cols[model_name]
        seg_mae <- compute_r_segment_mae(r_df, col, r_horizon)
        for (seg in names(seg_mae)) {
          results[[model_name]][[seg]] <- seg_mae[[seg]]
        }
      }
    }
    
    # Toto — single origin (window_id == 7, starts 2026-01-08)
    toto_seg_mae <- compute_toto_single_origin_mae(toto_horizon, target_window_id = 7)
    for (seg in names(toto_seg_mae)) {
      results$`Toto-2.5B`[[seg]] <- toto_seg_mae[[seg]]
    }
  }
  
  return(results)
}

plot_single_origin_comparison <- function(results) {
  # Reshape results to data.table
  plot_list <- list()
  for (model_name in names(results)) {
    m_res <- results[[model_name]]
    for (seg_name in names(m_res)) {
      val <- m_res[[seg_name]]
      if (!is.na(val)) {
        plot_list[[length(plot_list) + 1]] <- data.table(
          Model = model_name,
          Segment = seg_name,
          MAE = val
        )
      }
    }
  }
  
  if (length(plot_list) == 0) {
    cat("No single origin comparison data found.\n")
    return()
  }
  
  plot_df <- rbindlist(plot_list)
  plot_df[, Segment := factor(Segment, levels = names(SEGMENTS))]
  plot_df[, Model := factor(Model, levels = c("SARIMA", "MSTL+ARIMA", "Naive", "Toto-2.5B"))]
  
  colors_map <- c(
    "SARIMA" = "#e74c3c",
    "MSTL+ARIMA" = "#3498db",
    "Naive" = "#2ecc71",
    "Toto-2.5B" = "#9b59b6"
  )
  
  shapes_map <- c(
    "SARIMA" = 16,     # circle
    "MSTL+ARIMA" = 15, # square
    "Naive" = 17,      # triangle
    "Toto-2.5B" = 18   # diamond
  )
  
  # For annotation placement, we can use geom_text
  p2 <- ggplot(plot_df, aes(x = Segment, y = MAE, color = Model, group = Model, shape = Model)) +
    geom_line(size = 1.0) +
    geom_point(size = 3.5) +
    geom_text(aes(label = sprintf("%.0f", MAE)), vjust = -0.7, size = 3, fontface = "bold", show.legend = FALSE) +
    scale_color_manual(values = colors_map) +
    scale_shape_manual(values = shapes_map) +
    labs(
      title = "Error Growth by Horizon Depth:\nAutoregressive (SARIMA, MSTL+ARIMA, Naive) vs CPM (Toto-2.5B)\nAll models fitted once on 8,760h context, no re-scoring",
      x = "Forecast Horizon Depth",
      y = "MAE (vehicles/hour)"
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(face = "bold", size = 11, hjust = 0.5),
      legend.position = "bottom"
    )
  
  ggsave(file.path(PLOTS_DIR, "horizon_comparison.png"), plot = p2, width = 9, height = 5.5, dpi = 150)
  cat("  Saved: horizon_comparison.png\n")
}

print_comparison_table <- function(results) {
  seg_display <- c(
    "0-167h\n(1 week)" = "1wk",
    "168-335h\n(2 weeks)" = "2wk",
    "336-503h\n(3 weeks)" = "3wk",
    "504-671h\n(4 weeks)" = "4wk"
  )
  
  cat("\nMAE by horizon segment (single origin, 2026-01-08):\n")
  cat(sprintf("%-15s", "Model"))
  for (seg in names(seg_display)) {
    cat(sprintf("%10s", seg_display[seg]))
  }
  cat("\n")
  
  for (model in c("Toto-2.5B", "SARIMA", "MSTL+ARIMA", "Naive")) {
    cat(sprintf("%-15s", model))
    for (seg in names(seg_display)) {
      val <- results[[model]][[seg]]
      if (!is.null(val) && !is.na(val)) {
        cat(sprintf("%10.1f", val))
      } else {
        cat(sprintf("%10s", "N/A"))
      }
    }
    cat("\n")
  }
}

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

main <- function() {
  cat("============================================================\n")
  cat("1. CPM Stress-Test (Toto-only, all rolling windows)\n")
  cat("============================================================\n")
  cpm_results <- plot_cpm_stress_test()
  
  for (horizon_str in names(cpm_results)) {
    cat(sprintf("\n  Horizon=%sh:\n", horizon_str))
    h_res <- cpm_results[[horizon_str]]
    for (seg in names(h_res)) {
      val <- h_res[[seg]]
      seg_display <- gsub("\n", " ", seg)
      if (!is.na(val)) {
        cat(sprintf("    %-20s: MAE=%.1f\n", seg_display, val))
      } else {
        cat(sprintf("    %-20s: N/A\n", seg_display))
      }
    }
  }
  
  cat("\n============================================================\n")
  cat("2. Single-Origin Comparison (R + Toto, 2026-01-08 origin)\n")
  cat("============================================================\n")
  comparison_results <- compute_all_single_origin()
  plot_single_origin_comparison(comparison_results)
  print_comparison_table(comparison_results)
}

main()
